api/client/websocketupgradeproto/managedconn.go

```diff
@@ -0,0 +1,468 @@
+// Teleport
+// Copyright (C) 2023  Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package websocketupgradeproto
+
+import (
+	"io"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+)
+
+// 128 KB buffers were chosen as a resonable default for network I/O.
+// A Teleport Proxy behind a L7 load balancer will receive every single
+// connection through a WebSocket upgrade, so having reasonably sized buffers
+// helps with performance without using too much memory.
+// No buffer size will grow beyond this value, backpressure will be applied
+// by blocking writes until the reader has consumed some data or the connection
+// can be written to again.
```

Comment:
The proxy handles _a lot_ of connections, I'd feel better about a buffer size similar to the one of a TCP connection (8KB by default) rather than sixteen times as much buffer, unless the overhead of websocket framing is actually that bad.

Reply:
(if we bring it down to 8K I wouldn't even bother with the buffer resizing and just set `initialBufferSize` to 8K as well)

Reply:
A smaller write chunk size would also make it easier to close things on time and reply to pings more promptly.